### ISSUE TITLE
Usability for `MultiArch`

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#2349]
Some observations and questions:

* Should we have a default `ranks`? Maybe `ranks = (1, MPI.Comm_size(MPI.COMM_WORLD), 1)` or something? Or is this a bad idea? I forsee boilerplate for scripts that are intended to automatically scale...
* It'd be nice if we didn't have to pass the topology to _both_ `MultiArch` and `RectilinearGrid`; the grid constructor for `MultiArch` can grab topology from `arch`.
* Probably `MultiArch` should take a similar approach as the grid constructor; ie we can't distribute a grid in `Flat` directions, so `ranks` should be a 2-tuple for 2D topology, 1-element for 1D topology.
* We might want some functions for `arch` (and also `grid`, `model`, `sim`):
    * `communicator(grid)` returns eg `MPI.COMM_WORLD`
    * `local_rank(grid)` returns eg `MPI.Comm_rank(communicator(grid))` (or maybe just `rank(grid)`, or something)
    * This is maybe not our scope (belongs more to `MPI.jl`) but it does seem like it'd be nice to have a macro that a) prints normally for non-mpi and b) prints just from rank 0 if using MPI. That way we only have to change `arch` when switching from single-process runs to distributed runs.

**Comment by glwagner** [comment]
Can someone explain the purpose of `MPI.Init()`? Is it a bad idea to call this inside `MultiArch` constructor rather than requiring users to write it out? Can we detect if it's been called in the constructor and then call _by default_ if it hasn't been called yet?

**Comment by glwagner** [comment]
And of course the big one is that I'd prefer `MultiArchitecture` as the name of this object, but that isn't a necessary change (wondering what others think...)

**Comment by johnryantaylor** [comment]
Just to follow on from what @francispoulin wrote, we normally use MPI_Init() at the start and MPI_Finalize() at the end.  All code that calls MPI needs to be between these statements.  To your question about whether you can call MPI_Init() within MultiArch (or MultiArchitecture), I think the question is whether it would ever be useful to use an MPI routines before calling MultiArch.  For example, you can't find out the local rank until you call MPI_Init().  Also, it is good practice to use both MPI_Init() and MPI_Finalize(), so if MPI_Init() is in MultiArch, it leaves the question of where MPI_Finalize() should be. This might be an argument for having them both outside of MultiArch where they are visible to the user.

**Comment by francispoulin** [comment]
I agree with @johnryantaylor , even though it might make for shorter code, there are good reasons for keeping `MPI_Init()` and `MPI_Finalize()`, in each script.  Maybe this should be considered good practice?

**Comment by glwagner** [comment]
It does make sense, so we have to live with that necessary distinction for now. But lets continue to think about how to bring the syntax for distributed and single process scripts as close together as possible.

**Comment by glwagner** [comment]
@simone-silvestri suggests removing `topology` as an argument from the `MultiArch` constructor. `MultiArch` now uses `topology` to construct `RankConnectivity` (ie the mapping between ranks). But in @simone-silvestri's words "connectivity is a grid thing".

I'm not 100% sure how to solve it though. One possibility is to 1) rebuild `MultiArch` in the grid constructor adding connectivity information. Or we can 2) introduce a wrapper for grids on distributed domains or 3) add `connectivity` to all grids, set to `nothing` when not distributred. I think 1) is easiest but maybe not the cleanest.

**Comment by glwagner** [comment]
After some discussions I also think "Distributed" is a confusing name for this module. The module is really specific to "multi-process" execution with MPI, rather than distributed-memory execution (though it _supports_ distributed memory setups). `MultiRegion` also has distributed memory features.

I think drawing a closer connection to MPI might help, something like:

1. `MultiArch` -> `MPIArchitecture`
2. `Distributed` -> `MPIArchitectures`

Or, if we want to be even more explicit, `MPIMultiProcess` or something. But I think "MPI" uniquely implies "multi-process".

**Comment by simone-silvestri** [comment]
(1l ) should be already implemented. 

I.e., grid.architecture is necessary equal to the input architecture

**Comment by simone-silvestri** [comment]
(1l ) should be already implemented. 

I.e., grid.architecture is necessary equal to the input architecture.

I think the way to go is specify a partitioning object. That will define how you split the grid. The connectivity itself goes in the BC

**Comment by glwagner** [comment]
I guess `partition` would take the place of `ranks`.

But we have to store this information in either `grid` or `grid.architecture` (regardless of how it's specified by the user). This could look like

```julia
arch = MPIArchitecture() # note no arguments
grid = RectilinearGrid(arch, grid_kw..., partition=XPartition())
```

internally, the constructors maybe take the form

```julia
function RectilinearGrid(arch, size, other_grid_kw, architecture_kwargs...)
    arch = rebuild_arch(arch; architecture_kwargs...)
    etc
```

That's possibility (1).

Possibility (2) is to use a wrapper for MPI jobs, perhaps completely eliminating `MPIArchitecture` altogether:

```julia
arch = CPU()
grid = RectilinearGrid(arch, kw...)
grid = MultiProcessGrid(grid, partition=XPartition())
```

Possibilty (3) is to change all the `struct` definitions for grids so they have additional information pertinent to multi-process stuff, which is set to `nothing` when unused.

The downside to (2) is that it introduces yet another level of indirection, on top of both `ImmersedBoundaryGrid` and `MultiRegionGrid`. If we wanted to do MPI-distribued, multi-region simulations in complex domains, we have 4 (!) levels of wrapping. Maybe it's better to avoid that?

**Comment by simone-silvestri** [comment]
We also have possibility (3) of having
```
arch = MPIArchitecture(partition = XPartition(Ranks)) 
grid = RectilinearGrid(arch, grid_kw...)
```

**Comment by glwagner** [comment]
> It will allow us embed a little more multi process in the code.

Can you explain this point? To me it looks like the main difference is the _user interface_ rather than functionality, because every grid has an `architecture`. What am I missing?

**Comment by glwagner** [comment]
I'll convert this to a discussion since there are no immediate action items.
